operator/controllers/remote.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/kyma-project/manifest-operator/operator/pkg/labels"
+	"github.com/kyma-project/manifest-operator/operator/pkg/manifest"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"path"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type RemoteInterface struct {
+	NativeClient client.Client
+	CustomClient client.Client
+	NativeObject *v1alpha1.Manifest
+}
+
+func NewRemoteInterface(ctx context.Context, nativeClient client.Client, nativeObject *v1alpha1.Manifest) (*RemoteInterface, *v1alpha1.Manifest, error) {
+	namespacedName := client.ObjectKeyFromObject(nativeObject)
+	kymaOwner, ok := nativeObject.Labels[labels.ComponentOwner]
+	if !ok {
+		return nil, nil, fmt.Errorf(
+			fmt.Sprintf("label %s not found on manifest resource %s", labels.ComponentOwner, namespacedName))
+	}
+
+	remoteCluster := &custom.ClusterClient{DefaultClient: nativeClient}
+	restConfig, err := remoteCluster.GetRestConfig(ctx, kymaOwner, nativeObject.Namespace)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	customClient, err := remoteCluster.GetNewClient(restConfig, client.Options{
+		Scheme: nativeClient.Scheme(),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	remoteInterface := &RemoteInterface{
+		NativeObject: nativeObject,
+		NativeClient: nativeClient,
+		CustomClient: customClient,
+	}
+
+	if meta.IsNoMatchError(err) {
+		if err := remoteInterface.CreateCRD(ctx); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// check for remote resource
+	remoteManifest, err := remoteInterface.GetLatestRemote(ctx)
+
+	if client.IgnoreNotFound(err) != nil {
+		// unknown error
+		return nil, nil, err
+	} else if err != nil {
+		// remote resource doesn't exist
+		remoteManifest, err = remoteInterface.createRemote(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// check finalizer
+	if !controllerutil.ContainsFinalizer(remoteManifest, labels.ManifestFinalizer) {
+		controllerutil.AddFinalizer(remoteManifest, labels.ManifestFinalizer)
+		if err = remoteInterface.Update(ctx, remoteManifest); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return remoteInterface, remoteManifest, nil
+}
+
+func (r *RemoteInterface) createRemote(ctx context.Context) (*v1alpha1.Manifest, error) {
+	remoteManifest := &v1alpha1.Manifest{}
+	remoteManifest.Name = r.NativeObject.Name
+	remoteManifest.Namespace = r.NativeObject.Namespace
+	remoteManifest.Spec = *r.NativeObject.Spec.DeepCopy()
+	controllerutil.AddFinalizer(remoteManifest, labels.ManifestFinalizer)
+	err := r.CustomClient.Create(ctx, remoteManifest)
+	return remoteManifest, err
+}
+
+func (r *RemoteInterface) SpecSyncToNative(ctx context.Context, remoteManifest *v1alpha1.Manifest) (bool, error) {
+	if remoteManifest.Status.ObservedGeneration == 0 {
```

Comment:
if the remote manifest is freshly installed, wouldn't you be on generation 1 here? Generation 0 doesn't exist unless you use the object that was created in the operator here. Even if you do, you assume you use an object that was not updated from the API.